### DISCUSSION TITLE
[Proposal]: Customizable equality comparison of records.

### DISCUSSION THREAD

**Discussion by FLAMESpl** [dotnet/csharplang#5707] [Language Ideas] [answered]
# Customizable equality comparison of records.

## Summary

As far as I know, there is no option to overwrite generated equality check of record types. When `Equals` method override is added to the record type then error appears "Member with the same signature is already declared". We could make use of some mechanism that allows for customization of generated equality check.

## Motivation

Records equality check is generated based on its members so it is possible to indirectly customize it by modifying equality comparison of properties' types. This way structural comparison of a record can be guaranteed by making its properties records too, however, it is not possible when property's type is in another assembly. Best example are system collection types such as Array, List, IEnumerable and others. Structural comparison of collections can have some variant such as order sensitiveness and duplicate sensitiveness so its best to decide for equality check implementation on use case basis.

## Proposed solution

Add an attribute for properties that will override default equality comparison. Attribute would have an generic parameter for `IEqualityComparer<>` to be used for comparison.

**Comment by 333fred** [comment, marked as answer]
You cannot implement `Equals(object)`, but you can implement `Equals(RecordType)`. That is the way to customize the equality of a record type.

**Comment by HaloFour** [comment]
The team has expressed that they are not interest in wading into the policy questions/concerns when it comes to customizing equality with records.  They instead left it possible to provide your own implementations of the generated members as a way to provide that customization, and to use tools like source generators if you want to drive the implementation by attributes.

**Reply by FLAMESpl** [reply]
Sufficient for me
